control_protocol.md

```diff
@@ -0,0 +1,359 @@
+4. A content header (abbreviated with "H" in examples).
+5. Message content: The optional payload, which can be 0 or more frames.
+
+
+#### Directory
+
+Each Coordinator shall have a list of the Components connected to it.
+This is its _local Directory_.
+
+The _global Directory_ is the combination of the local directories of all Coordinators in a Network.
+
+
+### Conversation protocol
+
+In the protocol examples, `CA`, `CB`, etc. indicate Component Names.
+`N1`, `N2`, etc. indicate Node Namespaces and `Co1`, `Co2` their corresponding Coordinators.
+
+Here the Message content is expressed in plain English, for the exact definition see {ref}`message-layer`.
+
+In the exchange of messages, only the messages over the wire are shown, the connection identity of the ROUTER socket is not shown.
+
+
+#### Communication with the Coordinator
+
+(sign-in)=
+##### Initial connection
+
+After connecting to a Coordinator (`Co1`), a Component (`CA`) shall send a SIGNIN message indicating its Component Name.
+The Coordinator shall respond with an ACKNOWLEDGE, giving the Namespace and other relevant information, or with an ERROR, if the Component Name is already taken.
+After that successful handshake, the Coordinator shall store the connection identity and corresponding Component Name in its local {ref}`directory`.
+It shall also publish to the other Coordinators in the network that this Component signed in, see {ref}`Coordinator coordination<coordinator-coordination>`.
+Similarly, the Component shall store the Namespace and use it from this moment to generate its full name.
+
+If a Component does send a message to someone without having signed in, the Coordinator shall refuse message handling and return an error.
+
+:::{mermaid}
+sequenceDiagram
+    Note over CA,N1: Name "CA" is still free
+    participant N1 as N1.COORDINATOR
+    CA ->> N1: V|COORDINATOR|CA|H|SIGNIN
+    Note right of N1: Connection identity "IA"
+    Note right of N1: Stores "CA" with identity "IA"
+    N1 ->> CA: V|N1.CA|N1.COORDINATOR|H|ACKNOWLEDGE: Namespace is "N1"
+    Note left of CA: Stores "N1" as Namespace
+    Note over CA,N1: Name "CA" is already used
+    CA ->> N1: V|COORDINATOR|CA|H|SIGNIN
+    N1 ->> CA: V|CA|N1.COORDINATOR|H|ERROR: Name "CA" is already used.
+    Note left of CA: May retry with another Name
+    Note over CA,N1: "CA" has not send SIGNIN
+    Note left of CA: Wants to send a message to CB
+    CA ->> N1: V|N1.CB|CA|H|Content
+    Note right of N1: Does not know CA
+    N1 ->> CA: V|CA|N1.COORDINATOR|H|ERROR:I do not know you
+    Note left of CA: Should send a SIGNIN message
+:::
+
+
+##### Heartbeat
+
+We use heartbeat to know, whether a communication partner is still online.
+
+Every message received counts as a heartbeat.
+
+A Component should and a Coordinator shall send a PING and wait some time before considering a connection dead.
+A Coordinator shall follow the {ref}`sign out routine<signing-out>` for a signed in Component considered dead.
+
```

Comment:
```suggestion

After not having received a heartbeat from a peer for a chosen _heartbeat interval_, a Component should and a Coordinator shall send a PING and wait for a PONG response for 3 heartbeat intervals before considering a connection dead.
A Coordinator shall follow the {ref}`sign out routine<signing-out>` for a signed-in Component considered dead.
A Component may reduce its heartbeat interval when it receives a PING. 

```
3 heartbeat intervals should (I think) cover the case of a reconnecting Coordinator -- all reconnection activity should start at the latest after 1 interval.

Reply:
If the Coordinator is restarting, then it does not know the Component, so they have to sign in again.

If a Component reconnects, it has a new Connection identity and has to sign in again.

The only use case for PING is, that a Component did not send any message.
The answer should arrive within one heartbeat interval (at least if it is around 1 s).

> A Component may reduce its heartbeat interval when it receives a PING. 

What is your idea behind this?

Reply:
> thee component will send a PING, which the Component answers.

That confused me.

Now I got it.

Reply:
Let us discuss in #40

Reply:
I marked it as TBD, so it will be handled later with its own issue.